pkg/engine/engine.go

```diff
@@ -450,14 +444,32 @@ func (e *Engine) Logs(ctx context.Context, id string, follow bool, cancel bool,
 		}
 	}
 
-	file, err := os.Open(path)
+	stop := make(chan struct{})
+	file, err := newTailReader(path, stop)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error when newTailReader, err: %w", err)
 	}
 	defer file.Close()
-	reader := bufio.NewReader(file)
 
-	var prevBytes []byte
+	go func() {
+		for {
+			e.signalsLk.RLock()
+			_, running := e.signals[id]
+			e.signalsLk.RUnlock()
+			if !running {
+				time.Sleep(2 * time.Second)
+				close(stop)
+				return
+			}
+
+			time.Sleep(1 * time.Second)
+		}
+	}()
```

Comment:
This might be racy, and I think we can get rid of the polling. Can't we get the signal channel outside the goroutine, and then receive from it inside the goroutine?

Reply:
This might be running before the signal is even created, so we would need another channel to propagate signals... basically a subscription for all signals that govern the lifecycles of the tasks.

Why do you think it is racy? It might be delayed a few seconds, but that's fine.